### PR TITLE
Use IEditorFormattingService to format on commit

### PR DESCRIPTION
**Pull request by rchande**
Previously, if the user committed completion with semicolon (or another
auto-format triggering character), the completion commit code would run,
followed by the autoformat code. Both features run in places where
multiple edits happen sequentially, so they use
AutomaticCodeChangeMergePolicy when creating undo transactions. However,
this meant that commit followed by format-on-typechar would only produce
one undo transaction.

To avoid this, when committing, completion queries the relevant
IEditorFormattingService to determine if the commit character should
result in formatting, and then uses IEditorFormattingService to format
within a separate undo transaction with the default merge policy (the
command handler for format-on-typechar does essentially the same thing).

This results in two items in the undo stack: one for committing
completion, and one for the format. When the normal formatting command
handler runs, the formatter doesn't produce any changes, so it does
nothing.

Mind taking a look? @balajikris @basoundr @davkean @dpoeschl @jasonmalinowski @jmarolf @Pilchie

### CODE REVIEW

**Comment by rchande** [comment]
I'm pretty sure this is the right change, but if someone could actually try it out, that would be great.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs [hunk 1]
```diff
@@ -120,15 +121,17 @@ private void Commit(CompletionItem item, TextChange textChange, Model model, cha
...
 
-            if (item.ShouldFormatOnCommit)
+            var document = this.SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            var formattingService = document.GetLanguageService<IEditorFormattingService>();
+            if (formattingService != null && 
+                (item.ShouldFormatOnCommit || (commitChar != null && formattingService.SupportsFormattingOnTypedCharacter(document, commitChar.GetValueOrDefault()))))
             {
                 // Formatting the completion item affected span is done as a separate transaction because this gives the user
                 // the flexibility to undo the formatting but retain the changes associated with the completion item
                 using (var formattingTransaction = _undoHistoryRegistry.GetHistory(this.TextView.TextBuffer).CreateTransaction(EditorFeaturesResources.IntellisenseCommitFormatting))
```
Comment (Pilchie) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs]: Does IntelliSense need it's own transaction at this point?  If so, should it have a different string entry?

Comment (rchande) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs]: If you look at the top of this file (thanks github) you'll see that we use a different string for IntelliSense (EditorFeaturesResources.IntelliSense).

Comment (Pilchie) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs]: Ah, great.

**Comment by Pilchie** [comment]
:+1:

**Comment by rchande** [comment]
@basoundr (not sure why your comment doesn't show up inline) we always usethe EditorFormattingService now.

**Comment by rchande** [comment]
test this please

**Comment by basoundr** [comment]
@rchande Weird, the comment disappeared. I didn't notice the formatting service was used to format. You are right.

:+1:

**Pull request merged**